src/mat/impls/inv/matinv.c

```diff
@@ -66,35 +66,34 @@ static PetscErrorCode MatInvComputeNullSpace_Inv(Mat imat)
   MPI_Comm blockComm;
 
   PetscFunctionBeginI;
-  TRY( MatInvGetKSP(imat,&ksp) );
-  TRY( KSPGetPC(ksp,&pc) );
-  TRY( PCGetOperators(pc,&Kl,NULL) );
+  PetscValidHeaderSpecific(K,MAT_CLASSID,1);
+  if (pc) PetscValidHeaderSpecific(pc,PC_CLASSID,2);
+  TRY( PetscObjectTypeCompare((PetscObject)K,MATBLOCKDIAG,&blockdiag) );
+  if (blockdiag) {
```

Comment:
Having a blockdiag specific implementation calling MatComputeNullSpace_Default() on individual blocks would simplify the implementation.

Reply:
OK, I will try

Reply:
Actually, ability to do this speaks for the concept of this PR ;-)

Reply:
Done in cf5e155.